### ISSUE TITLE
Inconsistency requiring explicit casts on nullable value checked for null

### ISSUE THREAD

**Issue by andy-hanson** [AssemblyScript/assemblyscript#229]
```ts
class Point {
	constructor(readonly x: f64, readonly y: f64) { this.x = x; this.y = y; }
	public sum(): f64 {
		return this.x + this.y;
	}
}

function maybe_point(x: f64, y: f64): Point | null {
	return x + y === 0 ? null : new Point(x, y);
}

function sum(p: Point): f64 {
	return p.x + p.y;
}

export function test(a: f64, b: f64): f64 {
	let p = maybe_point(a, b);
	let s0 = p !== null ? p.sum() : 0;
	let s1 = p !== null ? sum(p) : 0;
	return s0 + s1;
}
```

There is no error when compiling with TypeScript. But when compiling with `asc` there is an error:
```
ERROR AS200: Conversion from type 'Point | null' to 'Point' requires an explicit cast.

 	let s1 = p !== null ? sum(p) : 0;
```

Ideally there would be no such errors since TypeScript can check that `p` is non-null at that location. This works for `s0` so I would expect it to also work for `s1`.

**Comment by dcodeIO** [comment]
Yeah, null checking in flows is still todo :)

**Comment by MaxGraey** [comment]
Closed in favour of #456
